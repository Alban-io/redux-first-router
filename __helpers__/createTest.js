import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (routesMap, actions, options = {}) => {
  if (actions && !Array.isArray(actions)) {
    options = actions
    actions = undefined
  }

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  expect(options).toMatchSnapshot('options')

  const routes = {
    FIRST: '/first',
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routesMap
  }

  const path = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { middleware, reducer, firstRoute, rudy } = createRouter(
    routes,
    options
  )

  const title = (state, action = {}) => {
    return action.payload
      ? action.type + '_' + JSON.stringify(action.payload)
      : action.type
  }
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  if (routesMap.FIRST) {
    expect(store.getState()).toMatchSnapshot('initial_state')
  }

  const firstAction = firstRoute()
  const res = await store.dispatch(firstAction)

  if (routesMap.FIRST) {
    expect(firstAction).toMatchSnapshot('first_action')
    expect(res).toMatchSnapshot('first_response')
    expect(store.getState()).toMatchSnapshot('first_state')
    expect(rudy.history.entries).toMatchSnapshot('first_history_entries')
    expect(rudy.history.index).toMatchSnapshot('first_history_index')
    expect(document.title).toMatchSnapshot('first_title')
  }

  for (const item of actions) {
    let res
    let postfix

    if (typeof item === 'string' && item.charAt(0) === '/') {
      res = await rudy.history.push(item)
      postfix = item
    }
    else {
      const act = typeof item === 'object' ? item : { type: item }
      const { type, params, query, hash, state, snipes } = act
      const action = { type }

      if (params) action.params = params
      if (query) action.query = query
      if (hash) action.hash = hash
      if (state) action.state = state

      res = await store.dispatch(action)
      postfix = '_' + type

      expect(action).toMatchSnapshot('action' + postfix)
    }

    expect(res).toMatchSnapshot('response' + postfix)
    expect(store.getState()).toMatchSnapshot('state' + postfix)
    expect(rudy.history.entries).toMatchSnapshot('history_entries' + postfix)
    expect(rudy.history.index).toMatchSnapshot('history_index' + postfix)
    expect(document.title).toMatchSnapshot('title' + postfix)
  }

  const types = Object.keys(routesMap)

  for (const type of types) {
    const route = routes[type]
    const postfix = '_' + type

    if (route.beforeLeave) {
      expect(route.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave' + postfix)
    }

    if (route.beforeEnter && route.beforeEnter.mock) { // allow redirect shortcut route options to work
      expect(route.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter' + postfix)
    }

    if (route.onLeave) {
      expect(route.onLeave.mock.calls.length).toMatchSnapshot('onLeave' + postfix)
    }

    if (route.onEnter) {
      expect(route.onEnter.mock.calls.length).toMatchSnapshot('onEnter' + postfix)
    }

    if (route.thunk) {
      expect(route.thunk.mock.calls.length).toMatchSnapshot('thunk' + postfix)
    }

    if (route.onComplete) {
      expect(route.onComplete.mock.calls.length).toMatchSnapshot('onComplete' + postfix)
    }

    if (route.onError) {
      expect(route.onError.mock.calls.length).toMatchSnapshot('onError' + postfix)
    }
  }

  if (options.beforeLeave) {
    expect(options.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave_options')
  }

  if (options.beforeEnter) {
    expect(options.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter_options')
  }

  if (options.onLeave) {
    expect(options.onLeave.mock.calls.length).toMatchSnapshot('onLeave_options')
  }

  if (options.onEnter) {
    expect(options.onEnter.mock.calls.length).toMatchSnapshot('onEnter_options')
  }

  if (options.thunk) {
    expect(options.thunk.mock.calls.length).toMatchSnapshot('thunk_options')
  }

  if (options.onComplete) {
    expect(options.onComplete.mock.calls.length).toMatchSnapshot('onComplete_options')
  }

  if (options.onError) {
    expect(options.onError.mock.calls.length).toMatchSnapshot('onError_options')
  }

  return {
    store,
    firstRoute,
    history: rudy.history,
    routes,
    location: () => store.getState().location
  }
}
