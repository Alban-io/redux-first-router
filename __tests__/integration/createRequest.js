import createTest from '../../__helpers__/createTest'

createTest('callback "bag" argument has all goodies', {
  SECOND: {
    path: '/second',
    beforeEnter: (bag) => expect(bag).toMatchSnapshot('bag')
  }
})
