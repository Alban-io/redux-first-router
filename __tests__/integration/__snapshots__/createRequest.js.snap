// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callback "bag" argument has all goodies: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {},
    "url": "/second",
  },
]
`;

exports[`callback "bag" argument has all goodies: 1 - history_index 1`] = `1`;

exports[`callback "bag" argument has all goodies: 1 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {},
        "url": "/second",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "SECOND",
}
`;

exports[`callback "bag" argument has all goodies: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`callback "bag" argument has all goodies: 1 - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`callback "bag" argument has all goodies: 1 - routes - SECOND - beforeEnter 1`] = `1`;

exports[`callback "bag" argument has all goodies: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {},
        "url": "/second",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`callback "bag" argument has all goodies: 1 - title 1`] = `"SECOND"`;

exports[`callback "bag" argument has all goodies: bag 1`] = `
Object {
  "_dispatched": false,
  "action": Object {
    "basename": "",
    "hash": "",
    "location": Object {
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/second",
          "search": "",
          "state": Object {},
          "url": "/second",
        },
      ],
      "from": null,
      "index": 1,
      "kind": "push",
      "length": 2,
      "pathname": "/second",
      "prev": Object {
        "basename": "",
        "entries": Array [
          Object {
            "basename": "",
            "hash": "",
            "key": "345678",
            "pathname": "/first",
            "search": "",
            "state": Object {},
            "url": "/first",
          },
        ],
        "hash": "",
        "index": 0,
        "kind": "load",
        "length": 1,
        "params": Object {},
        "pathname": "/first",
        "query": Object {},
        "scene": "",
        "search": "",
        "state": Object {},
        "status": 200,
        "type": "FIRST",
        "url": "/first",
      },
      "scene": "",
      "search": "",
      "status": 200,
      "url": "/second",
    },
    "params": Object {},
    "query": Object {},
    "state": Object {},
    "type": "SECOND",
  },
  "arg": "extra-arg",
  "cache": Object {
    "cacheAction": [Function],
    "clear": [Function],
    "isCached": [Function],
  },
  "commitDispatch": [Function],
  "commitHistory": [Function],
  "completed": false,
  "confirm": [Function],
  "ctx": Object {
    "busy": true,
  },
  "dispatch": [Function],
  "error": null,
  "getLocation": [Function],
  "getState": [Function],
  "getTitle": [Function],
  "hash": "",
  "history": MemoryHistory {
    "_forceRefresh": false,
    "_listener": [Function],
    "basename": "",
    "basenames": Array [],
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "firstRoute": Object {
      "commit": [Function],
      "nextHistory": Object {
        "basename": "",
        "basenames": Array [],
        "entries": Array [
          Object {
            "basename": "",
            "hash": "",
            "key": "345678",
            "pathname": "/first",
            "search": "",
            "state": Object {},
            "url": "/first",
          },
        ],
        "index": 0,
        "kind": "load",
        "length": 1,
        "location": Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        "saveHistory": [Function],
        "type": "@@rudy/UPDATE_HISTORY",
      },
      "type": "@@rudy/UPDATE_HISTORY",
    },
    "index": 0,
    "kind": "load",
    "length": 1,
    "location": Object {
      "basename": "",
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
    "saveHistory": [Function],
  },
  "initialLocation": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 0,
    "kind": "load",
    "length": 1,
    "params": Object {},
    "pathname": "/first",
    "prev": Object {
      "basename": "",
      "entries": Array [],
      "hash": "",
      "index": -1,
      "kind": "",
      "length": 0,
      "params": Object {},
      "pathname": "",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "type": "",
      "url": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "url": "/first",
  },
  "initialState": Object {
    "location": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "from": null,
      "hasSSR": false,
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "prev": Object {
        "basename": "",
        "entries": Array [],
        "hash": "",
        "index": -1,
        "kind": "",
        "length": 0,
        "params": Object {},
        "pathname": "",
        "query": Object {},
        "scene": "",
        "search": "",
        "state": Object {},
        "type": "",
        "url": "",
      },
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "title": "FIRST",
  },
  "options": Object {
    "callbacks": Array [
      "thunk",
      "onComplete",
      "beforeLeave",
      "beforeEnter",
      "onLeave",
      "onEnter",
      "thunk",
      "onComplete",
      "onError",
    ],
    "createRequest": [Function],
    "extra": Object {
      "arg": "extra-arg",
    },
    "initialEntries": Array [
      "/first",
    ],
    "onError": [Function],
    "shouldCall": [Function],
    "shouldTransition": [Function],
  },
  "params": Object {},
  "prevRoute": Object {
    "path": "/first",
    "type": "FIRST",
  },
  "query": Object {},
  "route": Object {
    "beforeEnter": [MockFunction] {
      "calls": Array [
        Array [
          [Circular],
        ],
      ],
    },
    "path": "/second",
    "type": "SECOND",
  },
  "routes": Object {
    "@@rudy/ADD_ROUTES": Object {
      "thunk": [Function],
      "type": "@@rudy/ADD_ROUTES",
    },
    "@@rudy/CALL_HISTORY": Object {
      "thunk": [Function],
      "type": "@@rudy/CALL_HISTORY",
    },
    "@@rudy/NOT_FOUND": Object {
      "path": "/not-found",
      "type": "@@rudy/NOT_FOUND",
    },
    "FIRST": Object {
      "path": "/first",
      "type": "FIRST",
    },
    "NEVER_USED_PATHLESS": Object {
      "thunk": [MockFunction],
      "type": "NEVER_USED_PATHLESS",
    },
    "REDIRECTED": Object {
      "onComplete": [MockFunction],
      "path": "/redirected",
      "type": "REDIRECTED",
    },
    "SECOND": Object {
      "beforeEnter": [MockFunction] {
        "calls": Array [
          Array [
            [Circular],
          ],
        ],
      },
      "path": "/second",
      "type": "SECOND",
    },
  },
  "state": Object {},
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
  },
  "tmp": Object {
    "from": Object {
      "basename": "",
      "hash": "",
      "location": Object {
        "entries": Array [
          Object {
            "basename": "",
            "hash": "",
            "key": "345678",
            "pathname": "/first",
            "search": "",
            "state": Object {},
            "url": "/first",
          },
          Object {
            "basename": "",
            "hash": "",
            "key": "345678",
            "pathname": "/second",
            "search": "",
            "state": Object {},
            "url": "/second",
          },
        ],
        "from": null,
        "index": 1,
        "kind": "push",
        "length": 2,
        "pathname": "/second",
        "prev": Object {
          "basename": "",
          "entries": Array [
            Object {
              "basename": "",
              "hash": "",
              "key": "345678",
              "pathname": "/first",
              "search": "",
              "state": Object {},
              "url": "/first",
            },
          ],
          "hash": "",
          "index": 0,
          "kind": "load",
          "length": 1,
          "params": Object {},
          "pathname": "/first",
          "query": Object {},
          "scene": "",
          "search": "",
          "state": Object {},
          "status": 200,
          "type": "FIRST",
          "url": "/first",
        },
        "scene": "",
        "search": "",
        "status": 200,
        "url": "/second",
      },
      "params": Object {},
      "query": Object {},
      "state": Object {},
      "type": "SECOND",
    },
  },
  "type": "SECOND",
}
`;
